src/relay/pass/kind_check.cc

```diff
@@ -20,94 +20,156 @@ namespace tvm {
 namespace relay {
 
 using namespace tvm::runtime;
-using Kind = TypeVarNode::Kind;
 
-struct KindChecker : TypeVisitor {
-  bool valid;
+struct KindChecker : TypeFunctor<Kind(const Type&)> {
+  const Module& mod;
 
-  KindChecker() : valid(true) {}
+  explicit KindChecker(const Module& mod) : mod(mod) {}
 
-  // checks if t is an incomplete node of kind k or a type param of kind k
-  bool MatchKind(const Type& t, Kind k) {
-    if (const IncompleteTypeNode* tv = t.as<IncompleteTypeNode>()) {
-      return tv->kind == k;
-    }
-
-    if (const TypeVarNode* tp = t.as<TypeVarNode>()) {
-      return tp->kind == k;
-    }
+  Kind VisitType_(const IncompleteTypeNode* op) override {
+    return op->kind;
+  }
 
-    return false;
+  Kind VisitType_(const TypeVarNode* op) override {
+    return op->kind;
   }
 
-  bool IsTypeKind(const Type& t) {
-    if (MatchKind(t, Kind::kType)) {
-      return true;
-    }
+  Kind VisitType_(const GlobalTypeVarNode* op) override {
+    return op->kind;
+  }
 
-    return t.as_derived<BaseTensorTypeNode>() || t.as<TupleTypeNode>() || t.as<FuncTypeNode>();
+  Kind VisitType_(const TensorTypeNode* op) override {
+    return Kind::kType;
   }
 
-  void VisitType_(const TupleTypeNode* op) override {
+  Kind VisitType_(const TupleTypeNode* op) override {
     // tuples should only contain normal types
     for (const Type& t : op->fields) {
-      this->VisitType(t);
-      valid = valid && IsTypeKind(t);
-      if (!valid) {
-        return;
-      }
+      Kind k = this->VisitType(t);
+      CHECK(k == Kind::kType)
+        << "All types in tuple type must be of a type kind but "
+        << t << " in " << GetRef<TupleType>(op) << " is of kind " << k;
     }
+    return Kind::kType;
   }
 
-  void VisitType_(const FuncTypeNode* op) override {
+  Kind VisitType_(const FuncTypeNode* op) override {
     // Func types should only take normal types for arguments
     // and only return a normal type. They should also have
     // well-formed constraints
+    FuncType ft = GetRef<FuncType>(op);
     for (const Type& t : op->arg_types) {
-      this->VisitType(t);
-      valid = valid && IsTypeKind(t);
-      if (!valid) {
-        return;
-      }
+      Kind k = this->VisitType(t);
+      CHECK(k == Kind::kType)
+        << "Function parameters must be of the type kind but parameter "
```

Comment:
Kind checking in this file looks quite repetitive, can we have something like:

```cpp
#define CHECK_KIND(msg, type, func_type, expected) \
  kind = this->VisitType(type); \
  CHECK_EQ(kind, expected) << (msg) << (type) << " of " << (func_type) << " is kind of " << (kind);
```

Reply:
I'll write a wrapper with the error reporter